File: 기타/201031 프로그래머스 챌린지/3번.py
def solution(v):
    answer = [0, 0, 0]
    global gv
    gv = v

    for i in range(len(gv)):
        for j in range(len(gv)):
            if gv[i][j] < 3:
                answer[gv[i][j]] += 1
                checkVisit(i, j, len(gv))

    return answer


def checkVisit(i, j, length):
    global gv
    gv[i][j] += 3
    if i > 0 and gv[i][j] % 3 == gv[i-1][j] % 3 and gv[i-1][j] < 3:
        checkVisit(i-1, j, length)
    if j > 0 and gv[i][j] % 3 == gv[i][j-1] % 3 and gv[i][j-1] < 3:
        checkVisit(i, j-1, length)
    if i < len(gv)-1 and gv[i][j] % 3 == gv[i+1][j] % 3 and gv[i+1][j] < 3:
        checkVisit(i+1, j, length)
    if j < len(gv)-1 and gv[i][j] % 3 == gv[i][j+1] % 3 and gv[i][j+1] < 3:
        checkVisit(i, j+1, length)


print(solution([[0, 0, 1], [2, 2, 1], [0, 0, 0]]))

File: 기타/210403 카카오/3번/3번.js
function solution(n, passenger, train) {
  const graph = Array.from(Array(n + 1), () => new Array());
  const visit = Array.from(Array(n + 1), () => new Array(n + 1).fill(false));
  const result = new Array(n + 1).fill(0);

  for (let t of train) {
    graph[t[0]].push(t[1]);
    graph[t[1]].push(t[0]);
  }

  for (let i = 2; i <= n; i++) {
    dfs(1, i, 0);
  }

  function dfs(now, end, current) {
    if (visit[end][now]) {
      return;
    }

    visit[end][now] = true;

    if (graph[now].includes(end)) {
      result[end] = Math.max(current + passenger[now - 1] + passenger[end - 1], result[end]);
      return;
    }

    for (let idx = 0; idx < graph[now].length; idx++) {
      if (visit[end][graph[now][idx]] === false) {
        dfs(graph[now][idx], end, current + passenger[now - 1]);
        visit[end][graph[now][idx]] = false;
      }
    }
  }

  const answer = [-1, -1];

  for (let i = 2; i <= n; i++) {
    if (result[i] >= answer[1]) {
      answer[0] = i;
      answer[1] = result[i];
    }
  }

  return answer;
}

const nn = 6;
const passe = [1, 1, 1, 1, 1, 1];
const tra = [
  [1, 2],
  [1, 3],
  [1, 4],
  [3, 5],
  [3, 6],
];

console.log(solution(nn, passe, tra));
